File: rxhttp/src/main/java/com/senierr/http/internal/ProgressResponseBody.java
package com.senierr.http.internal;

import android.support.annotation.NonNull;
import android.support.annotation.Nullable;

import com.senierr.http.RxHttp;

import java.io.IOException;

import okhttp3.MediaType;
import okhttp3.ResponseBody;
import okio.Buffer;
import okio.BufferedSource;
import okio.ForwardingSource;
import okio.Okio;
import okio.Source;

/**
 * 封装进度回调的响应体
 *
 * @author zhouchunjie
 * @date 2017/9/9
 */
public final class ProgressResponseBody extends ResponseBody {

    private @NonNull ResponseBody delegate;
    private @Nullable BufferedSource bufferedSource;
    private @NonNull OnProgressListener listener;

    public ProgressResponseBody(@NonNull ResponseBody responseBody,
                                @NonNull OnProgressListener listener) {
        this.delegate = responseBody;
        this.listener = listener;
    }

    @Override
    public MediaType contentType() {
        return delegate.contentType();
    }

    @Override
    public long contentLength() {
        return delegate.contentLength();
    }

    @Override
    public BufferedSource source() {
        if (bufferedSource == null) {
            bufferedSource = Okio.buffer(new CountingSource(delegate.source()));
        }
        return bufferedSource;
    }

    private final class CountingSource extends ForwardingSource {

        private long totalBytesRead;
        private long contentLength;
        private long lastRefreshTime;

        private CountingSource(Source delegate) {
            super(delegate);
        }

        @Override
        public long read(Buffer sink, long byteCount) throws IOException {
            long bytesRead = super.read(sink, byteCount);
            if (contentLength <= 0) {
                contentLength = contentLength();
            }

            totalBytesRead += bytesRead != -1 ? bytesRead : 0;
            final long curTime = System.currentTimeMillis();
            if (curTime - lastRefreshTime >= RxHttp.REFRESH_MIN_INTERVAL || totalBytesRead == contentLength) {
                int percent;
                if (contentLength <= 0) {
                    percent = 100;
                } else {
                    percent = (int) (totalBytesRead * 100 / contentLength);
                }
                listener.onProgress(new Progress(Progress.TYPE_DOWNLOAD, contentLength, totalBytesRead, percent));
                lastRefreshTime = System.currentTimeMillis();
            }
            return bytesRead;
        }
    }
}

File: rxhttp/src/main/java/com/senierr/http/model/ProgressResponseBody.kt
package com.senierr.http.model

import com.senierr.http.listener.OnProgressListener
import okhttp3.MediaType
import okhttp3.ResponseBody
import okio.*

/**
 * 封装进度回调的响应体
 *
 * @author zhouchunjie
 * @date 2017/9/9
 */
class ProgressResponseBody(
        private val delegate: ResponseBody,
        private val listener: OnProgressListener
) : ResponseBody() {

    private var bufferedSource: BufferedSource? = null

    override fun contentType(): MediaType? {
        return delegate.contentType()
    }

    override fun contentLength(): Long {
        return delegate.contentLength()
    }

    override fun source(): BufferedSource {
        val source = bufferedSource ?: CountingSource(delegate.source()).buffer()
        bufferedSource = source
        return source
    }

    private inner class CountingSource constructor(delegate: Source) : ForwardingSource(delegate) {

        private var totalBytesRead: Long = 0
        private var contentLength: Long = 0
        private var lastRefreshTime: Long = 0

        override fun read(sink: Buffer, byteCount: Long): Long {
            val bytesRead = super.read(sink, byteCount)
            if (contentLength <= 0) {
                contentLength = contentLength()
            }

            totalBytesRead += if (bytesRead != -1L) bytesRead else 0
            val curTime = System.currentTimeMillis()
            if (curTime - lastRefreshTime >= OnProgressListener.REFRESH_MIN_INTERVAL || totalBytesRead == contentLength) {
                val percent = if (contentLength <= 0) 100 else (totalBytesRead * 100 / contentLength).toInt()
                listener.onProgress(contentLength, totalBytesRead, percent)
                lastRefreshTime = System.currentTimeMillis()
            }
            return bytesRead
        }
    }
}